File: climateGuardian_Final_Projects/games/renewable_energy.py
import pygame
import random
import sys

def run_game():
    """Run the Renewable Energy Puzzle game - FULLY WORKING VERSION"""
    pygame.init()
    WIDTH, HEIGHT = 600, 600
    GRID_SIZE = 5
    CELL = 80
    GRID_OFFSET_Y = 150  # Space for UI at top
    screen = pygame.display.set_mode((WIDTH, HEIGHT))
    pygame.display.set_caption("🌍 Renewable Energy Puzzle - ClimateGuardian AI")
    clock = pygame.time.Clock()

    # Fonts
    title_font = pygame.font.SysFont('Arial', 40, bold=True)
    button_font = pygame.font.SysFont('Arial', 24, bold=True)
    info_font = pygame.font.SysFont('Arial', 20)

    # 0=empty, 1=Solar (needs sunny), 2=Wind (needs windy), 3=Hydro (needs river)
    map_grid = [
        [0, 1, 0, 1, 0],
        [2, 0, 2, 0, 2],
        [0, 3, 0, 3, 0],
        [2, 0, 2, 0, 2],
        [0, 1, 0, 1, 0]
    ]

    # Button class for better interaction
    class Button:
        def __init__(self, x, y, width, height, color, hover_color, text, value):
            self.rect = pygame.Rect(x, y, width, height)
            self.color = color
            self.hover_color = hover_color
            self.text = text
            self.value = value
            self.is_hovered = False
            self.is_selected = False
        
        def draw(self, screen):
            # Choose color based on state
            if self.is_selected:
                # Draw selection border
                border_rect = self.rect.inflate(10, 10)
                pygame.draw.rect(screen, (255, 255, 255), border_rect, border_radius=12)
                pygame.draw.rect(screen, self.color, self.rect, border_radius=10)
                # Draw text with contrasting color
                text_surface = button_font.render(self.text, True, (255, 255, 255))
                text_rect = text_surface.get_rect(center=self.rect.center)
                screen.blit(text_surface, text_rect)
                # Draw thick border
                pygame.draw.rect(screen, (0, 0, 0), self.rect, 4, border_radius=10)
            elif self.is_hovered:
                pygame.draw.rect(screen, self.hover_color, self.rect, border_radius=10)
                text_surface = button_font.render(self.text, True, (255, 255, 255))
                text_rect = text_surface.get_rect(center=self.rect.center)
                screen.blit(text_surface, text_rect)
                pygame.draw.rect(screen, (100, 100, 100), self.rect, 3, border_radius=10)
            else:
                pygame.draw.rect(screen, self.color, self.rect, border_radius=10)
                text_surface = button_font.render(self.text, True, (255, 255, 255))
                text_rect = text_surface.get_rect(center=self.rect.center)
                screen.blit(text_surface, text_rect)
                pygame.draw.rect(screen, (50, 50, 50), self.rect, 3, border_radius=10)
        
        def check_hover(self, mouse_pos):
            self.is_hovered = self.rect.collidepoint(mouse_pos)
        
        def check_click(self, mouse_pos):
            return self.rect.collidepoint(mouse_pos)

    # Create buttons
    buttons = [
        Button(50, 70, 140, 60, (255, 193, 7), (255, 220, 100), "☀️ SOLAR", 1),
        Button(220, 70, 140, 60, (33, 150, 243), (100, 181, 246), "💨 WIND", 2),
        Button(390, 70, 140, 60, (0, 188, 212), (77, 208, 225), "💧 HYDRO", 3)
    ]

    selected = None
    emissions = 500
    correct_placements = 0
    wrong_placements = 0
    game_over = False

    # Particle effect for correct placement
    particles = []

    class Particle:
        def __init__(self, x, y, color):
            self.x = x
            self.y = y
            self.color = color
            self.size = 10
            self.lifetime = 30
            self.vel_x = random.uniform(-2, 2)
            self.vel_y = random.uniform(-3, -1)
        
        def update(self):
            self.x += self.vel_x
            self.y += self.vel_y
            self.lifetime -= 1
            self.size = max(0, self.size - 0.3)
        
        def draw(self, screen):
            if self.size > 0:
                pygame.draw.circle(screen, self.color, (int(self.x), int(self.y)), int(self.size))

    running = True

    # Main game loop
    while running:
        mouse_pos = pygame.mouse.get_pos()
        
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
            
            if event.type == pygame.MOUSEBUTTONDOWN and not game_over:
                # Check button clicks
                for button in buttons:
                    if button.check_click(mouse_pos):
                        selected = button.value
                        # Update selected state
                        for b in buttons:
                            b.is_selected = (b.value == selected)
                        print(f"Selected energy type: {selected}")  # Debug
                
                # Check grid clicks
                if mouse_pos[0] >= 100 and mouse_pos[1] >= GRID_OFFSET_Y:
                    grid_x = (mouse_pos[0] - 100) // CELL
                    grid_y = (mouse_pos[1] - GRID_OFFSET_Y) // CELL
                    
                    if 0 <= grid_x < GRID_SIZE and 0 <= grid_y < GRID_SIZE and selected is not None:
                        cell_value = map_grid[grid_y][grid_x]
                        print(f"Clicked cell [{grid_y}][{grid_x}] = {cell_value}, Selected = {selected}")  # Debug
                        
                        if cell_value == selected:
                            # CORRECT MATCH!
                            emissions -= 50
                            correct_placements += 1
                            map_grid[grid_y][grid_x] = 0  # Clear the cell
                            
                            # Add green particles
                            particle_x = 100 + grid_x * CELL + CELL // 2
                            particle_y = GRID_OFFSET_Y + grid_y * CELL + CELL // 2
                            for _ in range(15):
                                particles.append(Particle(particle_x, particle_y, (0, 255, 0)))
                            
                            print(f"✅ CORRECT! Emissions: {emissions}")
                            
                        elif cell_value != 0:
                            # WRONG MATCH!
                            emissions += 30
                            wrong_placements += 1
                            
                            # Add red particles
                            particle_x = 100 + grid_x * CELL + CELL // 2
                            particle_y = GRID_OFFSET_Y + grid_y * CELL + CELL // 2
                            for _ in range(8):
                                particles.append(Particle(particle_x, particle_y, (255, 0, 0)))
                            
                            print(f"❌ WRONG! Emissions: {emissions}")
        
        # Update button hover states
        for button in buttons:
            button.check_hover(mouse_pos)
        
        # Update particles
        for particle in particles[:]:
            particle.update()
            if particle.lifetime <= 0:
                particles.remove(particle)
        
        # ==================== DRAWING ====================
        
        # Background gradient
        for i in range(HEIGHT):
            color_value = int(135 + (i / HEIGHT) * 100)
            pygame.draw.line(screen, (color_value, 206, 235), (0, i), (WIDTH, i))
        
        # Title with shadow
        shadow = title_font.render("🌍 Save The Planet!", True, (0, 50, 0))
        screen.blit(shadow, (WIDTH // 2 - shadow.get_width() // 2 + 3, 18))
        title = title_font.render("🌍 Save The Planet!", True, (0, 150, 0))
        screen.blit(title, (WIDTH // 2 - title.get_width() // 2, 15))
        
        # Draw buttons
        for button in buttons:
            button.draw(screen)
        
        # Instructions
        if selected is None:
            instruction = info_font.render("👆 Select an energy source above!", True, (50, 50, 50))
            screen.blit(instruction, (WIDTH // 2 - instruction.get_width() // 2, 140))
        else:
            energy_names = {1: "Solar ☀️", 2: "Wind 💨", 3: "Hydro 💧"}
            instruction = info_font.render(f"👇 Click matching colored tiles!", True, (50, 50, 50))
            screen.blit(instruction, (WIDTH // 2 - instruction.get_width() // 2, 140))
        
        # Draw Grid with shadow
        shadow_offset = 5
        pygame.draw.rect(screen, (100, 100, 100), 
                         (100 + shadow_offset, GRID_OFFSET_Y + shadow_offset, 
                          GRID_SIZE * CELL, GRID_SIZE * CELL), border_radius=10)
        
        for row in range(GRID_SIZE):
            for col in range(GRID_SIZE):
                color = (240, 240, 240)  # Default gray
                icon = ""
                
                if map_grid[row][col] == 1: 
                    color = (255, 235, 59)   # Solar = Bright Yellow
                    icon = "☀️"
                elif map_grid[row][col] == 2: 
                    color = (33, 150, 243)   # Wind = Blue
                    icon = "💨"
                elif map_grid[row][col] == 3: 
                    color = (0, 188, 212)    # Hydro = Cyan
                    icon = "💧"
                
                cell_rect = pygame.Rect(100 + col * CELL, GRID_OFFSET_Y + row * CELL, CELL, CELL)
                pygame.draw.rect(screen, color, cell_rect, border_radius=5)
                pygame.draw.rect(screen, (80, 80, 80), cell_rect, 2, border_radius=5)
                
                # Draw icon
                if icon:
                    icon_surface = button_font.render(icon, True, (50, 50, 50))
                    icon_rect = icon_surface.get_rect(center=cell_rect.center)
                    screen.blit(icon_surface, icon_rect)
        
        # Draw particles
        for particle in particles:
            particle.draw(screen)
        
        # Stats Panel
        stats_y = GRID_OFFSET_Y + GRID_SIZE * CELL + 20
        
        # Emissions bar background
        pygame.draw.rect(screen, (200, 200, 200), (50, stats_y, 500, 30), border_radius=15)
        
        # Emissions bar fill
        emissions_width = max(0, int((emissions / 500) * 500))
        if emissions <= 100:
            bar_color = (0, 200, 0)  # Green
        elif emissions <= 300:
            bar_color = (255, 165, 0)  # Orange
        else:
            bar_color = (255, 0, 0)  # Red
        
        pygame.draw.rect(screen, bar_color, (50, stats_y, emissions_width, 30), border_radius=15)
        
        # Emissions text
        emissions_text = info_font.render(f"💨 Emissions: {emissions} tons CO2", True, (0, 0, 0))
        screen.blit(emissions_text, (WIDTH // 2 - emissions_text.get_width() // 2, stats_y + 5))
        
        # Score
        score_text = info_font.render(f"✅ Correct: {correct_placements}  ❌ Wrong: {wrong_placements}", True, (50, 50, 50))
        screen.blit(score_text, (WIDTH // 2 - score_text.get_width() // 2, stats_y + 40))
        
        # Win condition
        if emissions <= 0 and not game_over:
            game_over = True
            print("🎉 GAME WON!")
            
        if game_over:
            # Overlay
            overlay = pygame.Surface((WIDTH, HEIGHT))
            overlay.set_alpha(220)
            overlay.fill((0, 150, 0))
            screen.blit(overlay, (0, 0))
            
            # Win message
            win_text = title_font.render("🎉 NET-ZERO ACHIEVED! 🎉", True, (255, 255, 255))
            screen.blit(win_text, (WIDTH // 2 - win_text.get_width() // 2, HEIGHT // 2 - 80))
            
            # Final score
            final_score = correct_placements * 100 - wrong_placements * 30
            final_score_text = title_font.render(f"Score: {final_score} points", True, (255, 255, 255))
            screen.blit(final_score_text, (WIDTH // 2 - final_score_text.get_width() // 2, HEIGHT // 2 - 20))
            
            # Stats
            stats_text = info_font.render(f"Perfect Matches: {correct_placements} | Mistakes: {wrong_placements}", True, (255, 255, 255))
            screen.blit(stats_text, (WIDTH // 2 - stats_text.get_width() // 2, HEIGHT // 2 + 30))
            
            # Exit instruction
            exit_text = info_font.render("Press any key or close window to exit...", True, (255, 255, 255))
            screen.blit(exit_text, (WIDTH // 2 - exit_text.get_width() // 2, HEIGHT // 2 + 70))
            
            # Check for exit
            keys = pygame.key.get_pressed()
            if any(keys):
                pygame.time.wait(500)
                running = False
        
        pygame.display.flip()
        clock.tick(60)

    pygame.quit()

if __name__ == "__main__":
    run_game()

File: climateGuardian_Final_Projects/js/games/renewable-energy.js
/* ==========================================
   RENEWABLE ENERGY PUZZLE GAME
   ========================================== */

function startRenewableEnergy() {
    document.getElementById('gameContainer').classList.remove('hidden');
    gameCanvas = document.getElementById('gameCanvas');
    gameCanvas.width = 600;
    gameCanvas.height = 600;
    gameCtx = gameCanvas.getContext('2d');
    
    currentGame = 'renewable';
    runRenewableEnergy();
}

function runRenewableEnergy() {
    const GRID_SIZE = 5;
    const CELL = 80;
    const GRID_OFFSET_Y = 150;

    let mapGrid = [
        [0, 1, 0, 1, 0],
        [2, 0, 2, 0, 2],
        [0, 3, 0, 3, 0],
        [2, 0, 2, 0, 2],
        [0, 1, 0, 1, 0]
    ];

    let selected = null;
    let emissions = 500;
    let correctPlacements = 0;
    let wrongPlacements = 0;
    let gameOver = false;
    let particles = [];

    const buttons = [
        { x: 50, y: 70, w: 140, h: 60, color: '#FFC107', hoverColor: '#FFD54F', text: '☀️ SOLAR', value: 1 },
        { x: 220, y: 70, w: 140, h: 60, color: '#2196F3', hoverColor: '#64B5F6', text: '💨 WIND', value: 2 },
        { x: 390, y: 70, w: 140, h: 60, color: '#00BCD4', hoverColor: '#4DD0E1', text: '💧 HYDRO', value: 3 }
    ];

    // Particle class for visual effects
    class Particle {
        constructor(x, y, color) {
            this.x = x;
            this.y = y;
            this.color = color;
            this.size = 10;
            this.lifetime = 30;
            this.velocityY = -2;
        }
        
        update() {
            this.y += this.velocityY;
            this.lifetime -= 1;
            this.size = Math.max(0, this.size - 0.3);
        }
        
        draw(ctx) {
            if (this.size > 0) {
                ctx.fillStyle = this.color;
                ctx.beginPath();
                ctx.arc(this.x, this.y, this.size, 0, Math.PI * 2);
                ctx.fill();
            }
        }
    }

    // Remove any existing click listeners
    const newCanvas = gameCanvas.cloneNode(true);
    gameCanvas.parentNode.replaceChild(newCanvas, gameCanvas);
    gameCanvas = newCanvas;
    gameCtx = gameCanvas.getContext('2d');
    gameCanvas.width = 600;
    gameCanvas.height = 600;

    function handleClick(e) {
        if (gameOver) return;
        
        const rect = gameCanvas.getBoundingClientRect();
        const x = e.clientX - rect.left;
        const y = e.clientY - rect.top;

        // Check button clicks
        buttons.forEach(btn => {
            if (x >= btn.x && x <= btn.x + btn.w && y >= btn.y && y <= btn.y + btn.h) {
                selected = btn.value;
            }
        });

        // Check grid clicks
        const gridX = Math.floor((x - 100) / CELL);
        const gridY = Math.floor((y - GRID_OFFSET_Y) / CELL);

        if (gridX >= 0 && gridX < GRID_SIZE && gridY >= 0 && gridY < GRID_SIZE && selected) {
            if (mapGrid[gridY][gridX] === selected) {
                emissions -= 50;
                correctPlacements++;
                mapGrid[gridY][gridX] = 0;
                
                // Add green particles for correct placement
                for (let i = 0; i < 10; i++) {
                    particles.push(new Particle(
                        100 + gridX * CELL + CELL / 2 + (Math.random() - 0.5) * 20,
                        GRID_OFFSET_Y + gridY * CELL + CELL / 2 + (Math.random() - 0.5) * 20,
                        '#4CAF50'
                    ));
                }
            } else if (mapGrid[gridY][gridX] !== 0) {
                emissions += 30;
                wrongPlacements++;
                
                // Add red particles for wrong placement
                for (let i = 0; i < 5; i++) {
                    particles.push(new Particle(
                        100 + gridX * CELL + CELL / 2 + (Math.random() - 0.5) * 20,
                        GRID_OFFSET_Y + gridY * CELL + CELL / 2 + (Math.random() - 0.5) * 20,
                        '#F44336'
                    ));
                }
            }
        }
    }

    gameCanvas.addEventListener('click', handleClick);

    function drawGame() {
        if (currentGame !== 'renewable') {
            gameCanvas.removeEventListener('click', handleClick);
            return;
        }

        // Update particles
        particles = particles.filter(p => {
            p.update();
            return p.lifetime > 0;
        });

        // Gradient background
        const gradient = gameCtx.createLinearGradient(0, 0, 0, 600);
        gradient.addColorStop(0, '#87CEEB');
        gradient.addColorStop(1, '#E0F7FA');
        gameCtx.fillStyle = gradient;
        gameCtx.fillRect(0, 0, 600, 600);

        // Title with shadow
        gameCtx.shadowColor = 'rgba(0, 0, 0, 0.3)';
        gameCtx.shadowBlur = 5;
        gameCtx.shadowOffsetX = 2;
        gameCtx.shadowOffsetY = 2;
        gameCtx.fillStyle = '#006400';
        gameCtx.font = 'bold 32px Arial';
        gameCtx.fillText('🌍 Save The Planet!', 120, 40);
        gameCtx.shadowBlur = 0;
        gameCtx.shadowOffsetX = 0;
        gameCtx.shadowOffsetY = 0;

        // Buttons with hover effect
        buttons.forEach(btn => {
            // Draw button shadow
            gameCtx.fillStyle = 'rgba(0, 0, 0, 0.2)';
            gameCtx.beginPath();
            roundRect(gameCtx, btn.x + 3, btn.y + 3, btn.w, btn.h, 10);
            gameCtx.fill();
            
            // Draw button
            gameCtx.fillStyle = btn.value === selected ? '#fff' : btn.color;
            gameCtx.beginPath();
            roundRect(gameCtx, btn.x, btn.y, btn.w, btn.h, 10);
            gameCtx.fill();
            
            // Draw border
            gameCtx.strokeStyle = btn.value === selected ? '#000' : '#333';
            gameCtx.lineWidth = btn.value === selected ? 4 : 3;
            gameCtx.stroke();
            
            // Draw text
            gameCtx.fillStyle = btn.value === selected ? btn.color : '#fff';
            gameCtx.font = 'bold 18px Arial';
            gameCtx.fillText(btn.text, btn.x + 20, btn.y + 38);
        });

        // Instructions
        gameCtx.fillStyle = '#333';
        gameCtx.font = '16px Arial';
        if (selected === null) {
            gameCtx.fillText('👆 Select an energy source above!', 170, 135);
        } else {
            gameCtx.fillText('👇 Click matching colored tiles below!', 160, 135);
        }

        // Grid with shadow
        gameCtx.fillStyle = 'rgba(0, 0, 0, 0.1)';
        gameCtx.fillRect(105, GRID_OFFSET_Y + 5, GRID_SIZE * CELL, GRID_SIZE * CELL);

        for (let row = 0; row < GRID_SIZE; row++) {
            for (let col = 0; col < GRID_SIZE; col++) {
                let color = '#f0f0f0';
                let icon = '';

                if (mapGrid[row][col] === 1) {
                    color = '#FFEB3B';
                    icon = '☀️';
                } else if (mapGrid[row][col] === 2) {
                    color = '#2196F3';
                    icon = '💨';
                } else if (mapGrid[row][col] === 3) {
                    color = '#00BCD4';
                    icon = '💧';
                }

                const cellX = 100 + col * CELL;
                const cellY = GRID_OFFSET_Y + row * CELL;

                gameCtx.fillStyle = color;
                gameCtx.beginPath();
                roundRect(gameCtx, cellX, cellY, CELL, CELL, 5);
                gameCtx.fill();
                
                gameCtx.strokeStyle = '#505050';
                gameCtx.lineWidth = 2;
                gameCtx.stroke();

                if (icon) {
                    gameCtx.font = '32px Arial';
                    gameCtx.fillText(icon, cellX + 24, cellY + 50);
                }
            }
        }

        // Draw particles
        particles.forEach(p => p.draw(gameCtx));

        // Stats Panel
        const statsY = GRID_OFFSET_Y + GRID_SIZE * CELL + 20;
        
        // Bar background
        gameCtx.fillStyle = '#c8c8c8';
        gameCtx.beginPath();
        roundRect(gameCtx, 50, statsY, 500, 30, 15);
        gameCtx.fill();

        // Bar fill
        const emissionsWidth = Math.max(0, (emissions / 500) * 500);
        const barColor = emissions <= 100 ? '#4CAF50' : emissions <= 300 ? '#FFA500' : '#F44336';
        gameCtx.fillStyle = barColor;
        gameCtx.beginPath();
        roundRect(gameCtx, 50, statsY, emissionsWidth, 30, 15);
        gameCtx.fill();

        // Bar text
        gameCtx.fillStyle = '#000';
        gameCtx.font = 'bold 16px Arial';
        gameCtx.fillText(`💨 Emissions: ${emissions} tons CO2`, 180, statsY + 20);

        // Score
        gameCtx.fillStyle = '#333';
        gameCtx.fillText(`✅ Correct: ${correctPlacements}  ❌ Wrong: ${wrongPlacements}`, 150, statsY + 55);

        // Win condition
        if (emissions <= 0 && !gameOver) {
            gameOver = true;
            
            // Overlay
            gameCtx.fillStyle = 'rgba(0, 150, 0, 0.95)';
            gameCtx.fillRect(0, 0, 600, 600);
            
            gameCtx.fillStyle = '#fff';
            gameCtx.font = 'bold 40px Arial';
            gameCtx.shadowColor = 'rgba(0, 0, 0, 0.5)';
            gameCtx.shadowBlur = 10;
            gameCtx.fillText('🎉 NET-ZERO ACHIEVED! 🎉', 50, 280);
            
            gameCtx.font = 'bold 24px Arial';
            const finalScore = correctPlacements * 100 - wrongPlacements * 30;
            gameCtx.fillText(`Final Score: ${finalScore} points`, 160, 330);
            gameCtx.fillText(`Perfect Matches: ${correctPlacements}`, 160, 370);
            
            gameCtx.shadowBlur = 0;
            
            logActivity('Renewable Energy Victory', correctPlacements * 10);
            ecoScore += correctPlacements * 10;
            document.getElementById('ecoScore').textContent = ecoScore;
            
            setTimeout(() => {
                gameCanvas.removeEventListener('click', handleClick);
                closeGame();
            }, 5000);
            return;
        }

        animationId = requestAnimationFrame(drawGame);
    }

    // Helper function for rounded rectangles
    function roundRect(ctx, x, y, w, h, r) {
        if (w < 2 * r) r = w / 2;
        if (h < 2 * r) r = h / 2;
        ctx.beginPath();
        ctx.moveTo(x + r, y);
        ctx.arcTo(x + w, y, x + w, y + h, r);
        ctx.arcTo(x + w, y + h, x, y + h, r);
        ctx.arcTo(x, y + h, x, y, r);
        ctx.arcTo(x, y, x + w, y, r);
        ctx.closePath();
        return ctx;
    }

    drawGame();
}
